File: metrics.py
#!/usr/bin/env python3

"""
Prometheus metrics for the NixOS Infrastructure Operator.

Provides observability into operator health, performance, and resource states.
"""

import logging
from prometheus_client import Counter, Gauge, Histogram, Info

logger = logging.getLogger(__name__)

# Info metrics
operator_info = Info("nio_operator", "NixOS Infrastructure Operator information")

# Machine metrics
machines_total = Gauge(
    "nio_machines_total",
    "Total number of managed machines",
    ["namespace"],
)

machines_discoverable = Gauge(
    "nio_machines_discoverable",
    "Number of discoverable machines",
    ["namespace"],
)

machines_with_configuration = Gauge(
    "nio_machines_with_configuration",
    "Number of machines with applied configuration",
    ["namespace"],
)

# Configuration metrics
configurations_total = Gauge(
    "nio_configurations_total",
    "Total number of NixOS configurations",
    ["namespace"],
)

configurations_applied = Counter(
    "nio_configurations_applied_total",
    "Total number of successful configuration applications",
    ["namespace", "machine"],
)

configurations_failed = Counter(
    "nio_configurations_failed_total",
    "Total number of failed configuration applications",
    ["namespace", "machine", "reason"],
)

# Reconciliation metrics
reconcile_duration = Histogram(
    "nio_reconcile_duration_seconds",
    "Time spent reconciling configurations",
    ["namespace", "configuration"],
    buckets=(1, 5, 10, 30, 60, 120, 300, 600, 1800, 3600),  # Up to 1 hour
)

reconcile_errors = Counter(
    "nio_reconcile_errors_total",
    "Total number of reconciliation errors",
    ["namespace", "configuration", "error_type"],
)

# SSH connection metrics
ssh_connections_total = Counter(
    "nio_ssh_connections_total",
    "Total number of SSH connection attempts",
    ["namespace", "machine", "result"],
)

ssh_connection_duration = Histogram(
    "nio_ssh_connection_duration_seconds",
    "Time to establish SSH connections",
    ["namespace", "machine"],
    buckets=(0.1, 0.5, 1, 2, 5, 10, 30),
)

# Git operations metrics
git_clones_total = Counter(
    "nio_git_clones_total",
    "Total number of Git clone operations",
    ["namespace", "repository", "result"],
)

git_clone_duration = Histogram(
    "nio_git_clone_duration_seconds",
    "Time to clone Git repositories",
    ["namespace", "repository"],
    buckets=(1, 5, 10, 30, 60, 120, 300),
)

# NixOS operations metrics
nixos_builds_total = Counter(
    "nio_nixos_builds_total",
    "Total number of NixOS builds",
    ["namespace", "machine", "build_type", "result"],
)

nixos_build_duration = Histogram(
    "nio_nixos_build_duration_seconds",
    "Time to build and apply NixOS configurations",
    ["namespace", "machine", "build_type"],
    buckets=(60, 300, 600, 1200, 1800, 3600, 7200),  # Up to 2 hours
)

# Retry metrics
retries_total = Counter(
    "nio_retries_total",
    "Total number of operation retries",
    ["operation", "attempt"],
)

retries_exhausted = Counter(
    "nio_retries_exhausted_total",
    "Total number of operations that exhausted all retries",
    ["operation"],
)

# Error metrics
errors_total = Counter(
    "nio_errors_total",
    "Total number of errors by type",
    ["error_type", "component"],
)

# Validation metrics
validation_errors = Counter(
    "nio_validation_errors_total",
    "Total number of input validation errors",
    ["validation_type", "field"],
)


def init_metrics():
    """Initialize metrics with operator information"""
    operator_info.info(
        {
            "version": "0.2.0",
            "name": "nixos-infrastructure-operator",
            "component": "operator",
        }
    )
    logger.info("Prometheus metrics initialized")


# Helper functions for common metric operations
def record_reconcile_success(namespace: str, configuration: str, duration: float):
    """Record successful reconciliation"""
    reconcile_duration.labels(namespace=namespace, configuration=configuration).observe(duration)


def record_reconcile_error(namespace: str, configuration: str, error_type: str):
    """Record reconciliation error"""
    reconcile_errors.labels(
        namespace=namespace, configuration=configuration, error_type=error_type
    ).inc()


def record_ssh_connection(namespace: str, machine: str, success: bool, duration: float):
    """Record SSH connection attempt"""
    result = "success" if success else "failure"
    ssh_connections_total.labels(namespace=namespace, machine=machine, result=result).inc()
    if success:
        ssh_connection_duration.labels(namespace=namespace, machine=machine).observe(duration)


def record_git_clone(namespace: str, repository: str, success: bool, duration: float):
    """Record Git clone operation"""
    result = "success" if success else "failure"
    git_clones_total.labels(namespace=namespace, repository=repository, result=result).inc()
    if success:
        git_clone_duration.labels(namespace=namespace, repository=repository).observe(duration)


def record_nixos_build(
    namespace: str, machine: str, build_type: str, success: bool, duration: float
):
    """Record NixOS build operation"""
    result = "success" if success else "failure"
    nixos_builds_total.labels(
        namespace=namespace, machine=machine, build_type=build_type, result=result
    ).inc()
    if success:
        nixos_build_duration.labels(
            namespace=namespace, machine=machine, build_type=build_type
        ).observe(duration)

File: internal/metrics/metrics.go
/*
Copyright 2026.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package metrics provides Prometheus metrics for the NixOS operator.
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"sigs.k8s.io/controller-runtime/pkg/metrics"
)

const (
	namespace = "nio"

	// ResultSuccess is the label value for successful operations.
	ResultSuccess = "success"
	// ResultFailure is the label value for failed operations.
	ResultFailure = "failure"
)

var (
	// Gauge metrics - current state

	// MachinesTotal is the total number of Machine resources.
	MachinesTotal = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "machines_total",
		Help:      "Total number of Machine resources",
	})

	// MachinesDiscoverable is the number of discoverable machines.
	MachinesDiscoverable = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "machines_discoverable",
		Help:      "Number of machines that are reachable via SSH",
	})

	// MachinesWithConfiguration is the number of machines with applied configuration.
	MachinesWithConfiguration = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "machines_with_configuration",
		Help:      "Number of machines with applied NixOS configuration",
	})

	// ConfigurationsTotal is the total number of NixosConfiguration resources.
	ConfigurationsTotal = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "configurations_total",
		Help:      "Total number of NixosConfiguration resources",
	})

	// JobsActive is the number of currently active apply jobs.
	JobsActive = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "jobs_active",
		Help:      "Number of currently active apply jobs",
	})

	// MachinesByState shows distribution of machines by state.
	MachinesByState = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "machines_by_state",
		Help:      "Number of machines by state (discoverable, undiscoverable)",
	}, []string{"state"})

	// ConfigsByState shows distribution of configurations by state.
	ConfigsByState = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "configs_by_state",
		Help:      "Number of configurations by state (pending, applying, applied, failed)",
	}, []string{"state"})

	// Counter metrics - accumulated values

	// ConfigurationsAppliedTotal is the total number of successful configuration applies.
	ConfigurationsAppliedTotal = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "configurations_applied_total",
		Help:      "Total number of successful configuration applies",
	})

	// ConfigurationsFailedTotal is the total number of failed configuration applies.
	ConfigurationsFailedTotal = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "configurations_failed_total",
		Help:      "Total number of failed configuration applies",
	})

	// SSHConnectionsTotal is the total number of SSH connection attempts.
	SSHConnectionsTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "ssh_connections_total",
		Help:      "Total number of SSH connection attempts",
	}, []string{"result"}) // success, failure

	// GitClonesTotal is the total number of git clone operations.
	GitClonesTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "git_clones_total",
		Help:      "Total number of git clone operations",
	}, []string{"result"}) // success, failure

	// NixosBuildsTotal is the total number of NixOS build operations.
	NixosBuildsTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "nixos_builds_total",
		Help:      "Total number of NixOS build operations",
	}, []string{"operation", "result"}) // operation: rebuild/anywhere, result: success/failure

	// RetriesTotal is the total number of retry attempts.
	RetriesTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "retries_total",
		Help:      "Total number of retry attempts",
	}, []string{"resource_type"}) // machine, configuration

	// ErrorsTotal is the total number of errors by type.
	ErrorsTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "errors_total",
		Help:      "Total number of errors by type",
	}, []string{"error_type"}) // ssh, git, nix, k8s

	// JobsFailedTotal is the total number of failed jobs.
	JobsFailedTotal = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "jobs_failed_total",
		Help:      "Total number of failed jobs",
	})

	// SecretWatchTriggersTotal is the total number of secret watch triggers.
	SecretWatchTriggersTotal = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "secret_watch_triggers_total",
		Help:      "Total number of reconciliations triggered by secret changes",
	})

	// Histogram metrics - durations

	// ReconcileDuration is the duration of reconcile operations.
	ReconcileDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
		Namespace: namespace,
		Name:      "reconcile_duration_seconds",
		Help:      "Duration of reconcile operations in seconds",
		Buckets:   prometheus.ExponentialBuckets(0.01, 2, 10), // 10ms to ~10s
	}, []string{"controller", "result"})

	// SSHConnectionDuration is the duration of SSH connection operations.
	SSHConnectionDuration = prometheus.NewHistogram(prometheus.HistogramOpts{
		Namespace: namespace,
		Name:      "ssh_connection_duration_seconds",
		Help:      "Duration of SSH connection operations in seconds",
		Buckets:   prometheus.ExponentialBuckets(0.1, 2, 8), // 100ms to ~25s
	})

	// GitCloneDuration is the duration of git clone operations.
	GitCloneDuration = prometheus.NewHistogram(prometheus.HistogramOpts{
		Namespace: namespace,
		Name:      "git_clone_duration_seconds",
		Help:      "Duration of git clone operations in seconds",
		Buckets:   prometheus.ExponentialBuckets(1, 2, 8), // 1s to ~256s
	})

	// NixosBuildDuration is the duration of NixOS build operations.
	NixosBuildDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
		Namespace: namespace,
		Name:      "nixos_build_duration_seconds",
		Help:      "Duration of NixOS build operations in seconds",
		Buckets:   prometheus.ExponentialBuckets(10, 2, 10), // 10s to ~2.8 hours
	}, []string{"operation"}) // rebuild, anywhere

	// JobDuration is the duration of apply jobs.
	JobDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
		Namespace: namespace,
		Name:      "job_duration_seconds",
		Help:      "Duration of apply jobs in seconds",
		Buckets:   prometheus.ExponentialBuckets(10, 2, 10), // 10s to ~2.8 hours
	}, []string{"operation", "result"})
)

func init() {
	// Register all metrics with controller-runtime metrics registry
	metrics.Registry.MustRegister(
		// Gauges
		MachinesTotal,
		MachinesDiscoverable,
		MachinesWithConfiguration,
		ConfigurationsTotal,
		JobsActive,
		MachinesByState,
		ConfigsByState,

		// Counters
		ConfigurationsAppliedTotal,
		ConfigurationsFailedTotal,
		SSHConnectionsTotal,
		GitClonesTotal,
		NixosBuildsTotal,
		RetriesTotal,
		ErrorsTotal,
		JobsFailedTotal,
		SecretWatchTriggersTotal,

		// Histograms
		ReconcileDuration,
		SSHConnectionDuration,
		GitCloneDuration,
		NixosBuildDuration,
		JobDuration,
	)
}

// RecordSSHConnection records an SSH connection attempt.
func RecordSSHConnection(success bool, duration float64) {
	result := ResultSuccess
	if !success {
		result = ResultFailure
	}
	SSHConnectionsTotal.WithLabelValues(result).Inc()
	SSHConnectionDuration.Observe(duration)
}

// RecordGitClone records a git clone operation.
func RecordGitClone(success bool, duration float64) {
	result := ResultSuccess
	if !success {
		result = ResultFailure
	}
	GitClonesTotal.WithLabelValues(result).Inc()
	GitCloneDuration.Observe(duration)
}

// RecordNixosBuild records a NixOS build operation.
func RecordNixosBuild(operation string, success bool, duration float64) {
	result := ResultSuccess
	if !success {
		result = ResultFailure
	}
	NixosBuildsTotal.WithLabelValues(operation, result).Inc()
	NixosBuildDuration.WithLabelValues(operation).Observe(duration)
}

// RecordJobCompletion records job completion.
func RecordJobCompletion(operation string, success bool, duration float64) {
	result := ResultSuccess
	if !success {
		result = ResultFailure
		JobsFailedTotal.Inc()
	}
	JobDuration.WithLabelValues(operation, result).Observe(duration)

	if success {
		ConfigurationsAppliedTotal.Inc()
	} else {
		ConfigurationsFailedTotal.Inc()
	}
}

// RecordError records an error by type.
func RecordError(errorType string) {
	ErrorsTotal.WithLabelValues(errorType).Inc()
}

// UpdateMachineState updates machine state metrics.
func UpdateMachineState(total, discoverable, configured int) {
	MachinesTotal.Set(float64(total))
	MachinesDiscoverable.Set(float64(discoverable))
	MachinesWithConfiguration.Set(float64(configured))
	MachinesByState.WithLabelValues("discoverable").Set(float64(discoverable))
	MachinesByState.WithLabelValues("undiscoverable").Set(float64(total - discoverable))
}

// UpdateConfigState updates configuration state metrics.
func UpdateConfigState(total, pending, applying, applied, failed int) {
	ConfigurationsTotal.Set(float64(total))
	ConfigsByState.WithLabelValues("pending").Set(float64(pending))
	ConfigsByState.WithLabelValues("applying").Set(float64(applying))
	ConfigsByState.WithLabelValues("applied").Set(float64(applied))
	ConfigsByState.WithLabelValues("failed").Set(float64(failed))
}
